### ISSUE TITLE
x/image/font/sfnt not all tables begin with four byte boundries 

### ISSUE THREAD

**Issue by marguerite** [golang/go#46384]
### What version of Go are you using (`go version`)?

<pre>
$ go version
go version go1.15.3 linux/amd64
</pre>

### What operating system and processor architecture are you using (`go env`)?

<details><summary><code>go env</code> Output</summary><br><pre>
$ go env
GO111MODULE="on"
GOARCH="amd64"
GOBIN=""
GOCACHE="/home/zhou/.cache/go-build"
GOENV="/home/zhou/.config/go/env"
GOEXE=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOINSECURE=""
GOMODCACHE="/home/zhou/go/pkg/mod"
GONOPROXY=""
GONOSUMDB=""
GOOS="linux"
GOPATH="/home/zhou/go"
GOPRIVATE=""
GOPROXY="https://goproxy.cn,direct"
GOROOT="/usr/lib64/go/1.15"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/usr/lib64/go/1.15/pkg/tool/linux_amd64"
GCCGO="gccgo"
AR="ar"
CC="gcc"
CXX="g++"
CGO_ENABLED="1"
GOMOD="/home/zhou/test/go.mod"
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build508378831=/tmp/go-build -gno-record-gcc-switches"
</pre></details>

### What did you do?

https://play.golang.org/p/0_ESQiUzpHy

/usr/share/fonts/truetype/wqy-zenhei.ttc is "WenQuanYi Zen Hei", the primary Simplified Chinese font for many Linux distributions for many years before Noto Sans CJK. nowadays many distributions still provide it, and some of them still don't have Noto Sans CJK fonts and didn't switch their default SC font yet. BTW, fontconfig project set it as the default SC font for programs relying on it without tweak.

In one word, if you want to print something Chinese with golang/x/image/font infrastructure, "wqy-zenhei.ttc" may come to your mind definitely.

### What did you expect to see?

nil error

### What did you see instead?

sfnt: invalid table offset

I print the tag with:

    func tag2string(u uint32) string {
      s := strconv.FormatUint(uint64(u), 16)
      decoded, _ := hex.DecodeString(s)
      return string(decoded)
    }

And find a weird tag "BDF" and "FFTM" , after some research I found they are fontforge specific tables, and are non-standard extension for opentype specification.

https://fontforge.org/docs/techref/non-standard.html#non-standard-bdf

the offset for previous BDF tag is 956, the length is 845.
the offset for FFTM tag is 8659, the length is 28. 

    if o&3 != 0 || n&3 != 0 {
      fmt.Println(u32string(tag))
      fmt.Println(o)
      fmt.Println(n)
    }
    // We ignore the checksums, but "all tables must begin on four byte
    // boundries [sic]".
    /*if o&3 != 0 {
      return nil, 0, false, errInvalidTableOffset
    }*/

<pre>BDF 
956
845
FFTM
8659
28
GDEF
8703
32
GPOS
12311719
32
GSUB
12311751
248
OS/2
12674224
86
cmap
1801
6854
cvt 
8655
4
gasp
8687
16
glyf
8735
12302984
head
12311999
54
hhea
12312053
36
hmtx
12312089
179544
loca
12491633
180092
maxp
12671725
32
name
12671757
2467
post
12674310
451503
vhea
13125813
36
vmtx
13125849
179326
FFTM
8659
28
GDEF
8703
32
GPOS
12311719
32
GSUB
13312927
248
OS/2
13315955
86
cmap
13305175
7752
cvt 
8655
4
gasp
8687
16
glyf
8735
12302984
head
13313175
54
hhea
12312053
36
hmtx
12312089
179544
loca
12491633
180092
maxp
12671725
32
name
13313229
2726
post
13316041
451447
EBDT
13770180
4120074
EBLC
17890254
563396
FFTM
8659
28
GDEF
8703
32
GPOS
12311719
32
GSUB
12311751
248
OS/2
12674224
86
cmap
1801
6854
cvt 
8655
4
gasp
8687
16
glyf
8735
12302984
head
18453650
54
hhea
12312053
36
hmtx
12312089
179544
loca
12491633
180092
maxp
12671725
32
post
12674310
451503
vhea
13125813
36
vmtx
13125849
179326
</pre>

because BDF table contains a string table, which is raw ASCII bytes. so it is hard to controll the 4 byte boundaries.
but seems fontforge do the byte alignment by itself as a whole.

so, I think, as many old "bimap embedded"  trutype fonts exist, can we skip this boundary check? simple comment it out saves wqy-zenhei.ttc

**Comment by seankhliao** [comment]
cc @nigeltao

**Comment by marguerite** [comment]
If anybody likes I can help implementing the parseBDF and parseFFTM functions, but I think they’re useless for truetype rendering and no one cares them...so it’s enough if we unlimit the table boundary check

**Comment by nigeltao** [comment]
I don't think we need parseBDF or parseFFTM functions, but https://docs.microsoft.com/en-us/typography/opentype/spec/otff clearly says "All tables must begin on four-byte boundaries".

Two questions:
1. Do you know if `wqy-zenhei.ttc` is usable on Microsoft Windows?
2. Do you know where in the C freetype code they enforce (or don't enforce) four-byte boundaries?

**Comment by marguerite** [comment]
@nigeltao 

1. Sure wqy-zenhei supports Windows...but due to the popularity of simsum.ttc( which I doubt has the same problem since it’s also a bimap embedded truetype), it’s hard to see it installed
2. These C codes do the table identification job in freetype2:

https://github.com/aseprite/freetype2/blob/master/src/sfnt/sfobjs.c#L451-L469
